src/dev/run_prepush_hook.ts

```diff
@@ -0,0 +1,312 @@
+import { getTimeReporter } from '@kbn/ci-stats-reporter';
+
+import { File } from './file';
+import { checkFileCasing } from './preflight_check/check_file_casing';
+import { checkTypescriptFiles } from './preflight_check/run_tsc';
+import { runUnitTests } from './preflight_check/run_unit_tests';
+import { esLintFiles } from './preflight_check/run_eslint';
+
+const HASH_FOR_NULL_OR_DELETED_FILE = '00000000000';
+
+function getDefaults(taskName: string) {
+  const TASK_COL_LENGTH = 10;
+  return {
+    task: chalk.blue(taskName.padEnd(TASK_COL_LENGTH)),
+    filename: '',
+  };
+}
+
+function nonNullable<T>(value: T): value is NonNullable<T> {
+  return value != null;
+}
+
+run(
+  async ({ log, flags }) => {
+    const toolingLog = new ToolingLog({
+      level: 'info',
+      writeTo: process.stdout,
+    });
+
+    const runStartTime = Date.now();
+    const reportTime = getTimeReporter(toolingLog, 'scripts/run_prepush_hook');
+
+    process.env.IS_KIBANA_PREPUSH_HOOK = 'true';
+
+    const git = simpleGit(REPO_ROOT);
+
+    const checkTypes = ['i18n', 'tsc', 'eslint', 'jest', 'fileCasing'] as const;
+
+    const checks = checkTypes.reduce((acc, check) => {
+      acc[check] = {
+        files: [],
+      };
+
+      return acc;
+    }, {} as Record<typeof checkTypes[number], { files: Array<{ path: string; file: File }> }>);
+
+    const { current } = await git.branchLocal();
+
+    const commonAncestor = (await git.raw(['merge-base', 'origin/main', current])).trim();
+
+    const diff = await git.diff([`${commonAncestor}..${current}`]);
+
+    const filesChangedSummaryTable = new Table({
+      head: ['Files changed in this PR compared to upstream:', 'Status'],
+      chars: { mid: '', 'left-mid': '', 'mid-mid': '', 'right-mid': '' },
+      colWidths: [80, 10],
+      wordWrap: true,
+      wrapOnWordBoundary: false,
+    });
+
+    const diffedFiles = diff
+      .split('diff --git')
+      .filter(Boolean)
+      .map((file) => {
+        const path = String(file.split(' b/')[1]).split('\n')[0];
+        if (!existsSync(path)) return undefined;
+
+        const [firstHash, secondHash] = String(file.split('index ')[1]).split('\n')[0].split('..');
+        const hunk = file.split(/@@ -[0-9]+,[0-9]+ \+[0-9]+,[0-9]+ @@/g)[1];
+
+        const mode =
+          firstHash === HASH_FOR_NULL_OR_DELETED_FILE
+            ? 'new'
+            : secondHash === HASH_FOR_NULL_OR_DELETED_FILE
+            ? 'deleted'
+            : 'modified';
+
+        filesChangedSummaryTable.push([path, mode]);
+
+        return {
+          path,
+          hunk,
+          mode,
+          added: hunk?.split('\n').filter((line) => line.startsWith('+')),
+          removed: hunk?.split('\n').filter((line) => line.startsWith('-')),
+        };
+      })
+      .filter(nonNullable);
+
+    if (diffedFiles.length <= Number(flags['max-files'])) {
+      log.info(`${diffedFiles.length} files changed in this PR compared to upstream.\n`);
```

Comment:
Just small word changes to be a bit more specific:

```suggestion
      log.info(`${diffedFiles.length} files changed in this branch compared to origin/main.\n`);
```